src/wallet/wallet.cpp

```diff
@@ -2627,6 +2671,38 @@ void CWallet::IncrementNoteWitnesses(
         bool performOrchardWalletUpdates)
 {
     LOCK(cs_wallet);
+
+    // Read the block from disk if we don't already have it.
+    const CBlock* pblock {pblockIn};
+    CBlock block;
+    if (!pblock) {
+        ReadBlockFromDisk(block, pindex, consensus);
+        pblock = &block;
+    }
+
+    // Update the wallet's note maps for spentness changes.
+    for (const CTransaction& tx : pblock->vtx) {
+        if (!tx.vJoinSplit.empty() || !tx.vShieldedSpend.empty()) {
+            for (std::pair<const uint256, CWalletTx>& wtxItem : mapWallet) {
```

Comment:
Someday we should redesign wallet data structures so that we don't have to iterate over all wallet transactions so often. Today is not that day. (And we'll probably only do so for a completely new fullnode wallet.)

Reply:
Something that we may do (not for this PR) short of a full rewrite is allow a mode that prunes fully-spent transactions from the wallet.